### PR TITLE
JAX/Flax safety checker

### PR DESCRIPTION
**Pull request by pcuenca**
Docstrings and tests pending.

To test:

```Python
from transformers import CLIPFeatureExtractor
from diffusers.pipelines.stable_diffusion import FlaxStableDiffusionSafetyChecker

feature_extractor = CLIPFeatureExtractor.from_pretrained("openai/clip-vit-large-patch14")
safety_checker, params = FlaxStableDiffusionSafetyChecker.from_pretrained(
    "<model id>", dtype= jnp.bfloat16,
)

params = replicate(params)
unshard = lambda x: einops.rearrange(x, 'd b ... -> (d b) ...')

@jax.pmap
def get_safety_scores(features, params):
    special_cos_dist, cos_dist = safety_checker.apply({"params": params}, features)
    return (special_cos_dist, cos_dist)

def run_safety_checker(images):
    pil_images = [Image.fromarray(image) for image in images]
    jnp_images = jnp.array(images)
    jnp_images = shard(jnp_images)
    features = feature_extractor(pil_images, return_tensors="np").pixel_values
    features = jnp.transpose(features, (0, 2, 3, 1))
    features = shard(features)
    special_cos_dist, cos_dist = get_safety_scores(features, params)
    images, has_nsfw = safety_checker.apply(
        {"params": unreplicate(params)},
        unshard(special_cos_dist),
        unshard(cos_dist),
        images,
        method=safety_checker.filtered_with_scores
    )
    return images, has_nsfw

shape = (8, 512, 512, 3)
images = np.random.rand(*shape)
images = (images * 255).round().astype("uint8")
_, has_nsfw = run_safety_checker(images)
```

### CODE REVIEW

File: src/diffusers/pipelines/stable_diffusion/safety_checker_flax.py [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+    norm_emb_1 = jnp.divide(emb_1.T, jnp.clip(jnp.linalg.norm(emb_1, axis=1), a_min=eps)).T
+    norm_emb_2 = jnp.divide(emb_2.T, jnp.clip(jnp.linalg.norm(emb_2, axis=1), a_min=eps)).T
+    return jnp.matmul(norm_emb_1, norm_emb_2.T)
+
+
+@flax_register_to_config
+class FlaxStableDiffusionSafetyChecker(nn.Module, FlaxModelMixin, ConfigMixin):
+    projection_dim: int = 768
+    # CLIPVisionConfig fields
+    vision_config: dict = field(default_factory=dict)
```
Comment (pcuenca) [inline src/diffusers/pipelines/stable_diffusion/safety_checker_flax.py:27]: The config file that is currently used contains all the CLIP configuration options. We just retrieve the ones corresponding to the `CLIPVisionConfig` that we use.

File: src/diffusers/pipelines/stable_diffusion/safety_checker_flax.py [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+
+    def __call__(self, clip_input):
+        pooled_output = self.vision_model(clip_input)[1]
+        image_embeds = self.visual_projection(pooled_output)
+
+        special_cos_dist = jax_cosine_distance(image_embeds, self.special_care_embeds)
+        cos_dist = jax_cosine_distance(image_embeds, self.concept_embeds)
+        return special_cos_dist, cos_dist
+
+    def filtered_with_scores(self, special_cos_dist, cos_dist, images):
```
Comment (pcuenca) [inline src/diffusers/pipelines/stable_diffusion/safety_checker_flax.py:64]: This method is not meant to be used with `pmap`, but it's fast as it just checks the scores.

File: src/diffusers/pipelines/stable_diffusion/safety_checker_flax.py (hunk 1 above)
Comment (patrickvonplaten) [reply src/diffusers/pipelines/stable_diffusion/safety_checker_flax.py:27]: Does the saving & loading work correctly with it?
Think long-term what  would be better is to actually just write the parameters that we would people like to change in here (e.g. the size of the CLIP, ....) and then use CLIP default config parameters (see suggestion below)

File: src/diffusers/pipelines/stable_diffusion/safety_checker_flax.py [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+            self.vision_config["image_size"],
+            self.vision_config["num_channels"],
+        )
+        pixel_values = jax.random.normal(rng, input_shape)
+        params_rng, dropout_rng = jax.random.split(rng)
+        rngs = {"params": params_rng, "dropout": dropout_rng}
+        return self.init(rngs, pixel_values)["params"]
+
+    def setup(self):
+        clip_vision_config = CLIPVisionConfig(**self.vision_config)
```
Comment (patrickvonplaten) [inline src/diffusers/pipelines/stable_diffusion/safety_checker_flax.py:44]: As a clean-up after the release I think this would be better. @patil-suraj what do you think? 

```suggestion
        clip_vision_config = CLIPVisionConfig(num_layers=self.config.num_layers, ...)
```

**Review by patrickvonplaten — APPROVED** [review]
If saving & loading works correctly good to go for me!

File: src/diffusers/pipelines/stable_diffusion/safety_checker_flax.py (hunk 1 above)
Comment (pcuenca) [reply src/diffusers/pipelines/stable_diffusion/safety_checker_flax.py:27]: > Does the saving & loading work correctly with it?

Loading works (from CLIP config files as the one currently in the `fusing` repo). I did not test saving yet, will do now.

> Think long-term what would be better is to actually just write the parameters that we would people like to change 

I thought about this approach, but there are many parameters in the configuration file and wasn't sure which ones we wanted to expose. In addition, it could still be useful to reuse any CLIP configuration json and just retrieve the vision part, instead of reading the parameters from the root of the json, which I presume would be incompatible with existing configurations.

These are all the relevant keys in the current configuration file:

```json
  "vision_config": {
    "_name_or_path": "",
    "add_cross_attention": false,
    "architectures": null,
    "attention_dropout": 0.0,
    "bad_words_ids": null,
    "bos_token_id": null,
    "chunk_size_feed_forward": 0,
    "cross_attention_hidden_size": null,
    "decoder_start_token_id": null,
    "diversity_penalty": 0.0,
    "do_sample": false,
    "dropout": 0.0,
    "early_stopping": false,
    "encoder_no_repeat_ngram_size": 0,
    "eos_token_id": null,
    "exponential_decay_length_penalty": null,
    "finetuning_task": null,
    "forced_bos_token_id": null,
    "forced_eos_token_id": null,
    "hidden_act": "quick_gelu",
    "hidden_size": 1024,
    "id2label": {
      "0": "LABEL_0",
      "1": "LABEL_1"
    },
    "image_size": 224,
    "initializer_factor": 1.0,
    "initializer_range": 0.02,
    "intermediate_size": 4096,
    "is_decoder": false,
    "is_encoder_decoder": false,
    "label2id": {
      "LABEL_0": 0,
      "LABEL_1": 1
    },
    "layer_norm_eps": 1e-05,
    "length_penalty": 1.0,
    "max_length": 20,
    "min_length": 0,
    "model_type": "clip_vision_model",
    "no_repeat_ngram_size": 0,
    "num_attention_heads": 16,
    "num_beam_groups": 1,
    "num_beams": 1,
    "num_channels": 3,
    "num_hidden_layers": 24,
    "num_return_sequences": 1,
    "output_attentions": false,
    "output_hidden_states": false,
    "output_scores": false,
    "pad_token_id": null,
    "patch_size": 14,
    "prefix": null,
    "problem_type": null,
    "pruned_heads": {},
    "remove_invalid_values": false,
    "repetition_penalty": 1.0,
    "return_dict": true,
    "return_dict_in_generate": false,
    "sep_token_id": null,
    "task_specific_params": null,
    "temperature": 1.0,
    "tf_legacy_loss": false,
    "tie_encoder_decoder": false,
    "tie_word_embeddings": true,
    "tokenizer_class": null,
    "top_k": 50,
    "top_p": 1.0,
    "torch_dtype": null,
    "torchscript": false,
    "transformers_version": "4.22.0.dev0",
    "typical_p": 1.0,
    "use_bfloat16": false
  },
  "vision_config_dict": {
    "hidden_size": 1024,
    "intermediate_size": 4096,
    "num_attention_heads": 16,
    "num_hidden_layers": 24,
    "patch_size": 14
  }
```

Comment (pcuenca) [reply src/diffusers/pipelines/stable_diffusion/safety_checker_flax.py:27]: > I did not test saving yet, will do now.

Saving works after #565 was applied.

File: src/diffusers/pipelines/stable_diffusion/safety_checker_flax.py (hunk 3 above)
Comment (patil-suraj) [reply src/diffusers/pipelines/stable_diffusion/safety_checker_flax.py:44]: The model is hardcoded here, so I think there is no advantage making it configurable like this. I'm okay with current approach.

File: src/diffusers/pipelines/stable_diffusion/safety_checker_flax.py [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+        self.vision_model = FlaxCLIPVisionModule(clip_vision_config)
```
Comment (patil-suraj) [inline src/diffusers/pipelines/stable_diffusion/safety_checker_flax.py:45]:
```suggestion
        self.vision_model = FlaxCLIPVisionModule(clip_vision_config, dtype=self.dtype)
```

File: src/diffusers/pipelines/stable_diffusion/safety_checker_flax.py [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+        self.visual_projection = nn.Dense(self.projection_dim, use_bias=False)
```
Comment (patil-suraj) [inline src/diffusers/pipelines/stable_diffusion/safety_checker_flax.py:46]:
```suggestion
        self.visual_projection = nn.Dense(self.projection_dim, use_bias=False, dtype=self.dtype)
```

**Review by patil-suraj — APPROVED** [review]
LGTM, thanks for adding this!

File: src/diffusers/pipelines/stable_diffusion/safety_checker_flax.py [hunk 6]
```diff
@@ -0,0 +1,111 @@
...
+            # at the cost of increasing the possibility of filtering benign image inputs
+            adjustment = 0.0
+
+            for concept_idx in range(len(special_cos_dist[0])):
+                concept_cos = special_cos_dist[i][concept_idx]
+                concept_threshold = self.special_care_embeds_weights[concept_idx].item()
+                result_img["special_scores"][concept_idx] = round(concept_cos - concept_threshold + adjustment, 3)
+                if result_img["special_scores"][concept_idx] > 0:
+                    result_img["special_care"].append({concept_idx, result_img["special_scores"][concept_idx]})
+                    adjustment = 0.01
```
Comment (jonatanklosko) [inline src/diffusers/pipelines/stable_diffusion/safety_checker_flax.py:83]: Hey! Is there any reason to increase `adjustment` on the first "special care" concept, rather than after the loop, before we check "bad" concepts? Intuitively I understand that we use "special care" concepts to make the "bad" concepts check more strict, but I feel like "special care" should be orthogonal to other "special care" detection.

In the current version it doesn't really matter if there's one or more special care concepts, so I think either approach would be the same, but I'm curious if there's a specific reason behind this.

This is relevant because the conditional `adjustment = 0.01` introduces a dependency between loop iterations. The alternative approach would be to move this outside the loop as: "if any special care score > 0 then set adjustment = 0.01 for bad concepts". With this, I think both loops could be vectorized (and in flax version they could probably be a part of `__call__`).

cc @patrickvonplaten

Comment (patrickvonplaten) [reply src/diffusers/pipelines/stable_diffusion/safety_checker_flax.py:83]: That's a good point - we can definitely move it out of the loop :-)

Comment (patrickvonplaten) [reply src/diffusers/pipelines/stable_diffusion/safety_checker_flax.py:83]: If you want feel free to open a PR for it :-)

Comment (patil-suraj) [reply src/diffusers/pipelines/stable_diffusion/safety_checker_flax.py:83]: Really good point @jonatanklosko

Comment (jonatanklosko) [reply src/diffusers/pipelines/stable_diffusion/safety_checker_flax.py:83]: Perfect, thanks! I will submit a PR sometime this week :)

Comment (patil-suraj) [reply src/diffusers/pipelines/stable_diffusion/safety_checker_flax.py:83]: awesome, thanks a lot!
